apps/ejabberd/src/ejabberd_router.erl

```diff
@@ -295,50 +310,62 @@ code_change(_OldVsn, State, _Extra) ->
 %%% Internal functions
 %%--------------------------------------------------------------------
 
-do_route(OrigFrom, OrigTo, OrigPacket) ->
-    ?DEBUG("route~n\tfrom ~p~n\tto ~p~n\tpacket ~p~n",
-           [OrigFrom, OrigTo, OrigPacket]),
-    %% Filter globally
-    case ejabberd_hooks:run_fold(filter_packet,
-                                 {OrigFrom, OrigTo, OrigPacket}, []) of
-        {From, To, Packet} ->
-            LDstDomain = To#jid.lserver,
-            case mnesia:dirty_read(external_component, LDstDomain) of
-                [] ->
-                    case mnesia:dirty_read(route, LDstDomain) of
-                        [] ->
-                            ejabberd_s2s:route(From, To, Packet);
-                        [#route{handler=Handler}] ->
-                            do_local_route(OrigFrom, OrigTo, OrigPacket,
-                                           LDstDomain, Handler)
-                    end;
-                [#external_component{handler = Handler}] ->
-                    do_local_route(OrigFrom, OrigTo, OrigPacket,
-                                   LDstDomain, Handler)
-            end;
-        drop ->
-            ejabberd_hooks:run(xmpp_stanza_dropped,
-                               OrigFrom#jid.lserver,
-                               [OrigFrom, OrigTo, OrigPacket]),
-            ok
-    end.
-
-do_local_route(OrigFrom, OrigTo, OrigPacket, LDstDomain, Handler) ->
-    %% Filter locally
-    case ejabberd_hooks:run_fold(filter_local_packet, LDstDomain,
-                                 {OrigFrom, OrigTo, OrigPacket}, []) of
-        {From, To, Packet} ->
-            case Handler of
-                {apply_fun, Fun} ->
-                    Fun(From, To, Packet);
-                {apply, Module, Function} ->
-                    Module:Function(From, To, Packet)
-            end;
+routing_modules_list() ->
+    %% this is going to be compiled on startup from settings
+    mod_routing_machine:get_routing_module_list().
+
+compile_routing_module() ->
+    Mods = ejabberd_config:get_local_option(routing_modules),
+    CodeStr = make_routing_module_source(Mods),
+    {Mod, Code} = dynamic_compile:from_string(CodeStr),
+    code:load_binary(Mod, "mod_routing_machine.erl", Code).
+
+make_routing_module_source(undefined) ->
+    ModList = [ejabberd_router_global,
+               ejabberd_router_external,
+               ejabberd_router_localdomain,
+               ejabberd_s2s],
+    make_routing_module_source(ModList);
+make_routing_module_source(Mods) ->
+    binary_to_list(iolist_to_binary(io_lib:format(
+        "-module(mod_routing_machine).~n"
+        "-compile(export_all).~n"
+        "get_routing_module_list() -> ~p.~n",
+        [Mods]))).
+
+route(_, _, _, []) ->
+    ok; %% shouldn't we raise error here? If we get here means we don't know
```

Comment:
Not exactly. When we start routing a message we have in the default setup four modules - global, external, localdomain and s2s, and each of them has a chance to drop or route the message. If we are here means no module knew what to do, every one passed it on. So, we have no idea what to do with this message, which doesn't seem good. Either we signal a problem, or install some 'catch-all' stuff, or just don't care. That's the choice.